### PR TITLE
Add ability to transmute (somewhat) with generic consts in arrays

### PR DESCRIPTION
**Pull request by JulianKnodt**
Previously if the expression contained generic consts and did not have a directly equivalent type, transmuting the type in this way was forbidden, despite the two sizes being identical. Instead, we should be able to lazily tell if the two consts are identical, and if so allow them to be transmuted.

This is done by normalizing the forms of expressions into sorted order of multiplied terms, which is not generic over all expressions, but should handle most cases.

This allows for some _basic_ transmutations between types that are equivalent in size without requiring additional stack space at runtime.

I only see one other location at which `SizeSkeleton` is being used, and it checks for equality so this shouldn't affect anywhere else that I can tell.

See [this Stackoverflow post](https://stackoverflow.com/questions/73085012/transmute-nested-const-generic-array-rust) for what was previously necessary to convert between types. This PR makes converting nested `T -> [T; 1]` transmutes possible, and `[uB*2; N] -> [uB; N * 2]` possible as well.

I'm not sure whether this is something that would be wanted, and if it is it definitely should not be insta-stable, so I'd add a feature gate.

### CODE REVIEW

File: src/test/ui/const-generics/transmute.rs
Comment (fbstj) [inline src/test/ui/const-generics/transmute.rs:11]:
```suggestion
fn ident<const W: usize, const H: usize>(v: [[u32; H]; W]) -> [[u32; H]; W] {
```
otherwise its the same as `transpose` above?

File: src/test/ui/const-generics/transmute.rs
Comment (fbstj) [inline src/test/ui/const-generics/transmute.rs:37]: is it worth having `flatten_inner` (`-> [[u32; D * W]; H]`) & `flatten_outer` (`-> [[u32; D]; H * W]` ?)

or is this intentionally mixing the levels to test something ese?

File: src/test/ui/const-generics/transmute.rs
Comment (fbstj) [inline src/test/ui/const-generics/transmute.rs:45]: is a `[u8; L * 2] -> [u16; L]` needed?

File: src/test/ui/const-generics/transmute.rs
Comment (JulianKnodt) [reply src/test/ui/const-generics/transmute.rs:37]: I don't think this is necessary, it's just testing that it compresses some number of layers without having to compress all of them.

**Comment by nagisa** [comment]
cc @eddyb may have some thoughts on the changes to the layout code.

File: compiler/rustc_middle/src/ty/layout.rs
Comment (eddyb) [inline compiler/rustc_middle/src/ty/layout.rs:402]: This doesn't imply "multiplication" to me, am I missing something?

File: compiler/rustc_middle/src/ty/layout.rs
Comment (eddyb) [inline compiler/rustc_middle/src/ty/layout.rs:421]: You probably want to use a `SmallVec`, given all the trouble you've went through to avoid calling `.sort()` (for performance, I assume?).

File: compiler/rustc_middle/src/ty/layout.rs
Comment (eddyb) [inline compiler/rustc_middle/src/ty/layout.rs:462]: Why would they be at the end? Is this a coincidence of what `<` actually does, or what?

File: compiler/rustc_middle/src/ty/layout.rs
Comment (eddyb) [inline compiler/rustc_middle/src/ty/layout.rs:396]: This doesn't need to be cheap, we could plausibly run a SAT/SMT solver here (since this is for "proving" two types have the same size). Feels weird to canonicalize ahead of time when `ty::Const` itself doesn't do it, but I don't know much about "generic consts".

File: compiler/rustc_hir_typeck/src/intrinsicck.rs
Comment (eddyb) [inline compiler/rustc_hir_typeck/src/intrinsicck.rs:90]: Is there no way to actually print the expression? "generic size != generic size" feels like very bad UX.

File: compiler/rustc_middle/src/ty/layout.rs
Comment (eddyb) [inline compiler/rustc_middle/src/ty/layout.rs:292]: This should be a `bug!` (`tcx.layout_of` call above should've succeeded if the length is constant, see the existing `bug!` just a few lines above).

File: compiler/rustc_middle/src/ty/layout.rs
Comment (eddyb) [inline compiler/rustc_middle/src/ty/layout.rs:303]: How come this doesn't call `tcx.expand_abstract_consts(len)`?

File: compiler/rustc_middle/src/ty/layout.rs
Comment (eddyb) [inline compiler/rustc_middle/src/ty/layout.rs:241]: This should definitely have its own separate documentation (including the fact that the constant has to be of type `usize` and represents the size of the type in bytes).

File: compiler/rustc_middle/src/ty/layout.rs
Comment (eddyb) [inline compiler/rustc_middle/src/ty/layout.rs:463]: This should be a `checked_mul` or go through `Size`, you don't want to let through overflows.

**Review by eddyb — CHANGES_REQUESTED** [review]
Overall the approach looks sound (if compiler/lang teams are fine with it etc.), left some comments, don't really have much to say other than nitpicking though.

File: compiler/rustc_middle/src/ty/layout.rs
Comment (JulianKnodt) [reply compiler/rustc_middle/src/ty/layout.rs:462]: This is a coincidence of the implementation of `Ord` on `Const`s if I remember correctly. I can make it independent of the ord implementation if that is better.

**Comment by JulianKnodt** [comment]
will update this after I get back from vacation in a week or so

File: compiler/rustc_middle/src/ty/layout.rs
Comment (JulianKnodt) [reply compiler/rustc_middle/src/ty/layout.rs:402]: I was thinking about almost like a merge in a merge-sort, I can rename it to better fit but since it was a private fn I didn't think too much about it

File: compiler/rustc_middle/src/ty/layout.rs
Comment (JulianKnodt) [reply compiler/rustc_middle/src/ty/layout.rs:401]: I'm not sure whether putting a whole SMT solver here would be reasonable, that seems like a lot of extra baggage for something that only gets rarely invoked.

File: compiler/rustc_hir_typeck/src/intrinsicck.rs
Comment (JulianKnodt) [reply compiler/rustc_hir_typeck/src/intrinsicck.rs:91]: For now, trying to print out an expression just prints `[const expr]`, which could be fixed but not sure if it should be tacked onto this pr

File: compiler/rustc_middle/src/ty/layout.rs
Comment (JulianKnodt) [reply compiler/rustc_middle/src/ty/layout.rs:297]: I think this can still be hit (I hit it while testing overflow), when the `layout_of` overflows for a nested array `[[u32; 99999999]; 99999999]`, but we do not overflow for just the outer array. Eventually we will get an overflow, but we don't want to throw a bug here

File: compiler/rustc_middle/src/ty/layout.rs
Comment (JulianKnodt) [reply compiler/rustc_middle/src/ty/layout.rs:308]: oops yes definitely should here

**Comment by nagisa** [comment]
r? rust-lang/compiler

**Comment by apiraino** [comment]
r? rust-lang/compiler

File: compiler/rustc_middle/src/ty/layout.rs
Comment (b-naber) [inline compiler/rustc_middle/src/ty/layout.rs:424]: This needs a more elaborate comment I feel given that this code isn't so straight forward. Can you maybe give an example and mention the intent behind this function?

File: compiler/rustc_middle/src/ty/layout.rs
Comment (b-naber) [inline compiler/rustc_middle/src/ty/layout.rs:451]: We should be able to do all of this in one loop instead of having three separate loops.

File: compiler/rustc_middle/src/ty/layout.rs
Comment (b-naber) [inline compiler/rustc_middle/src/ty/layout.rs:486]: This is not true, take e.g. `fn transpose<const W: usize, const H: usize>(v: [[u32; 2 * H]; W + W]) -> [[u32; W + W]; 2 * H]`.

Here in the `compute` call for `[[u32; 2 * H]; W + W]` we get the literal `8` because this statement holds for `[u32; 2*H]`, so we get `Mul(H, Mul(8, Add(W, W)))` for the whole term, whereas for `[[u32; W + W]; 2 * H]` we get `Mul(4, Add(W, W))` for the inner term `[u32; W + W]`, but then we can't simplify to the literal `8` because `Literal(n) < Binop(Add, W, W)`, so we end up with `Mul(H, Mul(2, Mul(4, Add(W, W))))`. 

I think you would have to either manually implement `Ord` or write a function that orders this solely for this functionality.

File: compiler/rustc_middle/src/ty/layout.rs
Comment (b-naber) [inline compiler/rustc_middle/src/ty/layout.rs:500]: I think having this function is not great. You should be able to only have one recursive function, that makes it easier to understand the code than having two separate recursive functions which call each other.

**Review by b-naber — CHANGES_REQUESTED** [review]
Probably need to take a second look at this in a couple of days, but have added some thoughts now.

File: compiler/rustc_hir_typeck/src/intrinsicck.rs
Comment (BoxyUwU) [reply compiler/rustc_hir_typeck/src/intrinsicck.rs:91]: you should just print the consts so that when printing for `ConstKind::Expr` functions correctly this code will print correctly. please dont implement correct formatting for `ConstKind::Expr` in this PR

File: compiler/rustc_middle/src/ty/layout.rs
Comment (b-naber) [inline compiler/rustc_middle/src/ty/layout.rs:496]: Why is this necessary? Don't we already have this multiplication in the `c.checked_mul(k)`?

File: compiler/rustc_middle/src/ty/layout.rs
Comment (b-naber) [inline compiler/rustc_middle/src/ty/layout.rs:488]: I don't think you need the for loop here.

**Review by b-naber — COMMENTED** [review]
Looks a lot better without the recursion 👍

This would be good to go after the nits were addressed imo (unless @BoxyUwU requests any changes).

I do think this might need lang-team approval though.

**Comment by b-naber** [comment]
Might need lang-team approval. This PR allows calling the `unsafe` function `std::mem::transmute` on types involving generic constants expressions, such as in:

```
fn transpose_with_const<const W: usize, const H: usize>(
  v: [[u32; 2 * H]; W + W]
) -> [[u32; W + W]; 2 * H]
```

**Comment by cramertj** [comment]
We discussed this in the @rust-lang/lang meeting today, and we were all excited about the idea of allowing more `transmute` calls. Many valid (but not provably so) `transmute` calls are currently rejected by the compiler's checks, pushing folks to less ergonomic options like `transmute_copy` or pointer casts.

We think the checks added in this change are valuable for allowing more correct code, and __we'd like to land this (feature-gated) change__.

The lang team would like to also __ensure that this is not stabilized without a larger rethinking of `transmute`'s size check__. There are many cases today in which a user knows that two types will be the same size, but `transmute` does not. Several of us expressed that __we want `transmute`'s existing size check moved to a lint__. Folks could `allow` when needed, and we could introduce more heuristics of the kind added here (heuristics are something that we're more wary of for hard errors).

N.B.: the current documentation for `transmute` today does say clearly `Both types must have the same size. Compilation will fail if this is not guaranteed.`. It may be necessary to introduce a new `transmute` function in order to ensure that no `unsafe` code has its expectations violated if it was using `transmute` to make size assertions.

**Comment by scottmcm** [comment]
For anyone looking to undertake such a rethink, I'll note that I recently made [a MIR `Transmute`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/enum.CastKind.html#variant.Transmute) which handles differently-sized types, so that rather than being malformed it instead is UB in both [CTFE](https://github.com/rust-lang/rust/pull/108442/files#r1145794713) and [codegen](https://github.com/rust-lang/rust/pull/108442/files#diff-980b4ebf35bb4374c16e69f1fda39ad6bba3174be1cc30c3378b1d0f18f351f9R27-R51).

So the implementation is there if you can figure out the best way to expose it 🙂

**Comment by b-naber** [comment]
LGTM

**Comment by b-naber** [comment]
@BoxyUwU Do you still plan to review this?

File: compiler/rustc_feature/src/active.rs
Comment (klensy) [inline compiler/rustc_feature/src/active.rs:520]: Maybe 1.70.0?

Comment (fee1-dead) [reply compiler/rustc_feature/src/active.rs:520]: Shouldn't this be CURRENT_RUSTC_VERSION?

Comment (klensy) [reply compiler/rustc_feature/src/active.rs:520]: maybe

Comment (klensy) [reply compiler/rustc_feature/src/active.rs]: Should be `CURRENT_RUSTC_VERSION` instead of explicit version.

**Comment by JulianKnodt** [comment]
Should be good now!

**Comment by b-naber** [comment]
Can anybody please explain why this needs `CURRENT_RUSTC_VERSION`? Not too familiar with the feature system... But I find it kind of weird that all other `active` features have a specific version number and the only other feature that uses  `CURRENT_RUSTC_VERSION` is an `incomplete` feature.

**Comment by b-naber** [comment]
nvm, just saw [@klensy's post](https://github.com/rust-lang/rust/pull/106281#discussion_r1160856864).

@bors r+ rollup

**Comment by scottmcm** [comment]
@b-naber The meta-reason is that it avoids the "oops, the fork happened, need to unapprove all the PRs with the version in them so they get updated with the new number" problem.  A new or stabilized gate always just uses the magic constant, and the release branching scripts make sure they get the real number.

**Pull request merged**
